Avoid re-doing CUDA stream sync in OwnerRRef

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59377 Move CUDA-related stuff of TP agent to separate file
* #59376 Make CUDA serde support for TP agent pluggable
* #59375 Merge TensorPipe's CPU and CUDA channel registry
* #59299 Remove LazyStreamContext (2 out of 2)
* #59298 Remove LazyStreamContext (1 out of 2)
* #59297 Fix CUDA sync when switching streams in RPC tests
* #58753 Fix race condition in TP agent
* #56863 Ensure async_execution works with CUDAFuture
* **#57355 Avoid re-doing CUDA stream sync in OwnerRRef**

We had started fixing OwnerRRef to make it CUDA-compatible, by properly synchronizing CUDA streams/events where appropriate. However, since we started using CUDAFuture (or, well, ivalue::Future nowadays, after they got merged) this is all done automatically for us, hence we can undo these "fixes" as they're now duplicated.

Differential Revision: [D28118182](https://our.internmc.facebook.com/intern/diff/D28118182/)

lw committed:
Avoid re-doing CUDA stream sync in OwnerRRef

We had started fixing OwnerRRef to make it CUDA-compatible, by properly synchronizing CUDA streams/events where appropriate. However, since we started using CUDAFuture (or, well, ivalue::Future nowadays, after they got merged) this is all done automatically for us, hence we can undo these "fixes" as they're now duplicated.

Differential Revision: [D28118182](https://our.internmc.facebook.com/intern/diff/D28118182/)

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -369,7 +369,6 @@ void RequestCallbackImpl::processPythonRemoteCall(
           IValue py_ivalue = jit::toIValue(result, PyObjectType::get());
 
           py::gil_scoped_release release;
-          ownerRRef->recordAllStreams(lsctx);
```
`recordAllStreams` recorded a new event in each of `lsctx`'s streams. However, in the line just below, we call `setValue` which calls `markCompleted` on the future, which does the same thing already.

mrshenli replied:
If we remove this, I think we will need to pass context devices to `getOrCreateOwnerRRef` on line 338, so that the created RRef will use CUDA ivalue::Future.

lw commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -415,7 +414,6 @@ void RequestCallbackImpl::processPythonRRefFetchCall(
       Message m =
           PythonRRefFetchRet(std::move(*result).toIValues()).toMessage();
       m.setId(messageId);
-      rref->blockAllStreams(lsctx);
```
`blockAllStreams` enqueues a wait for the RRef's events on each of `lsctx`'s streams. However, a few lines above, we call `getValue` on the RRef which calls `wait` on the future, which already does that.

lw commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -274,13 +265,7 @@ const IValue& OwnerRRef::getValue() const {
-  }
-  // Before accessing the value in this RRef, current CUDA streams must wait
-  // for pending CUDA operations that create the value.
-  blockCurrentStreams(events_);
```
`blockCurrentStreams` enqueues a wait for the RRef's events on each of the current streams, but just above we call `wait` on the future which does that too.

lw committed:
Update on "Avoid re-doing CUDA stream sync in OwnerRRef"


We had started fixing OwnerRRef to make it CUDA-compatible, by properly synchronizing CUDA streams/events where appropriate. However, since we started using CUDAFuture (or, well, ivalue::Future nowadays, after they got merged) this is all done automatically for us, hence we can undo these "fixes" as they're now duplicated.

Differential Revision: [D28118182](https://our.internmc.facebook.com/intern/diff/D28118182/)

[ghstack-poisoned]

lw committed:
Update on "Avoid re-doing CUDA stream sync in OwnerRRef"


We had started fixing OwnerRRef to make it CUDA-compatible, by properly synchronizing CUDA streams/events where appropriate. However, since we started using CUDAFuture (or, well, ivalue::Future nowadays, after they got merged) this is all done automatically for us, hence we can undo these "fixes" as they're now duplicated.

Differential Revision: [D28118182](https://our.internmc.facebook.com/intern/diff/D28118182/)

[ghstack-poisoned]

lw committed:
Update on "Avoid re-doing CUDA stream sync in OwnerRRef"


We had started fixing OwnerRRef to make it CUDA-compatible, by properly synchronizing CUDA streams/events where appropriate. However, since we started using CUDAFuture (or, well, ivalue::Future nowadays, after they got merged) this is all done automatically for us, hence we can undo these "fixes" as they're now duplicated.

Differential Revision: [D28118182](https://our.internmc.facebook.com/intern/diff/D28118182/)

[ghstack-poisoned]

mrshenli reviewed (commented):
Test failures are real

```
May 02 21:20:38 ======================================================================
May 02 21:20:38 ERROR [11.357s]: test_rref_to_here_synchronization4 (__main__.TensorPipeTensorPipeAgentCudaRpcTestWithSpawn)
May 02 21:20:38 ----------------------------------------------------------------------
May 02 21:20:38 Traceback (most recent call last):
May 02 21:20:38   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 374, in wrapper
May 02 21:20:38     self._join_processes(fn)
May 02 21:20:38   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 566, in _join_processes
May 02 21:20:38     self._check_return_codes(elapsed_time)
May 02 21:20:38   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 609, in _check_return_codes
May 02 21:20:38     raise RuntimeError(error)
May 02 21:20:38 RuntimeError: Process 1 exited with error code 10 and exception:
May 02 21:20:38 Traceback (most recent call last):
May 02 21:20:38   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 492, in run_test
May 02 21:20:38     getattr(self, test_name)()
May 02 21:20:38   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 376, in wrapper
May 02 21:20:38     fn()
May 02 21:20:38   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 93, in wrapper
May 02 21:20:38     return func(*args, **kwargs)
May 02 21:20:38   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/distributed/rpc/rpc_test.py", line 5616, in test_rref_to_here_synchronization4
May 02 21:20:38     self._test_rref_synchronization("cuda:0", "cuda:1")
May 02 21:20:38   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/distributed/rpc/rpc_test.py", line 5598, in _test_rref_synchronization
May 02 21:20:38     self.assertEqual(actual, expected)
May 02 21:20:38   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1352, in assertEqual
May 02 21:20:38     super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
May 02 21:20:38   File "/opt/conda/lib/python3.6/unittest/case.py", line 682, in assertTrue
May 02 21:20:38     raise self.failureException(msg)
May 02 21:20:38 AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1.3e-06 and atol=1e-05, found 2000 element(s) (out of 2000) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 0.2334456741809845 (0.0 vs. 0.2334456741809845), which occurred at index (19, 8).
May 02 21:20:38 
```

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -415,7 +414,6 @@ void RequestCallbackImpl::processPythonRRefFetchCall(
       Message m =
```
I cannot comment on line 412, so add comments here:

```
result = std::make_shared<SerializedPyObj>(
            pythonRpcHandler.serialize(jit::toPyObject(rref->getValue())));
```

The above code will only block current streams instead of the ones in `lsctx`. I guess we need to add a `MultiStreamGuard` to wrap `getValue()`?

lw committed:
Update on "Avoid re-doing CUDA stream sync in OwnerRRef"


We had started fixing OwnerRRef to make it CUDA-compatible, by properly synchronizing CUDA streams/events where appropriate. However, since we started using CUDAFuture (or, well, ivalue::Future nowadays, after they got merged) this is all done automatically for us, hence we can undo these "fixes" as they're now duplicated.

Differential Revision: [D28118182](https://our.internmc.facebook.com/intern/diff/D28118182/)

[ghstack-poisoned]

lw committed:
Update on "Avoid re-doing CUDA stream sync in OwnerRRef"


We had started fixing OwnerRRef to make it CUDA-compatible, by properly synchronizing CUDA streams/events where appropriate. However, since we started using CUDAFuture (or, well, ivalue::Future nowadays, after they got merged) this is all done automatically for us, hence we can undo these "fixes" as they're now duplicated.

Differential Revision: [D28118182](https://our.internmc.facebook.com/intern/diff/D28118182/)

[ghstack-poisoned]

lw committed:
Update on "Avoid re-doing CUDA stream sync in OwnerRRef"


We had started fixing OwnerRRef to make it CUDA-compatible, by properly synchronizing CUDA streams/events where appropriate. However, since we started using CUDAFuture (or, well, ivalue::Future nowadays, after they got merged) this is all done automatically for us, hence we can undo these "fixes" as they're now duplicated.

Differential Revision: [D28118182](https://our.internmc.facebook.com/intern/diff/D28118182/)

[ghstack-poisoned]

lw committed:
Update on "Avoid re-doing CUDA stream sync in OwnerRRef"


We had started fixing OwnerRRef to make it CUDA-compatible, by properly synchronizing CUDA streams/events where appropriate. However, since we started using CUDAFuture (or, well, ivalue::Future nowadays, after they got merged) this is all done automatically for us, hence we can undo these "fixes" as they're now duplicated.

Differential Revision: [D28118182](https://our.internmc.facebook.com/intern/diff/D28118182/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM. We might want to rerun one of @pbelevich's pipeline code to check the impact.

lw committed:
Update on "Avoid re-doing CUDA stream sync in OwnerRRef"


We had started fixing OwnerRRef to make it CUDA-compatible, by properly synchronizing CUDA streams/events where appropriate. However, since we started using CUDAFuture (or, well, ivalue::Future nowadays, after they got merged) this is all done automatically for us, hence we can undo these "fixes" as they're now duplicated.

Differential Revision: [D28118182](https://our.internmc.facebook.com/intern/diff/D28118182/)

[ghstack-poisoned]

lw commented:
> We might want to rerun one of @pbelevich's pipeline code to check the impact.

That won't work right now because it puts Locks inside RRefs.

lw committed:
Update on "Avoid re-doing CUDA stream sync in OwnerRRef"


We had started fixing OwnerRRef to make it CUDA-compatible, by properly synchronizing CUDA streams/events where appropriate. However, since we started using CUDAFuture (or, well, ivalue::Future nowadays, after they got merged) this is all done automatically for us, hence we can undo these "fixes" as they're now duplicated.

Differential Revision: [D28118182](https://our.internmc.facebook.com/intern/diff/D28118182/)

[ghstack-poisoned]

lw committed:
Update on "Avoid re-doing CUDA stream sync in OwnerRRef"


We had started fixing OwnerRRef to make it CUDA-compatible, by properly synchronizing CUDA streams/events where appropriate. However, since we started using CUDAFuture (or, well, ivalue::Future nowadays, after they got merged) this is all done automatically for us, hence we can undo these "fixes" as they're now duplicated.

Differential Revision: [D28118182](https://our.internmc.facebook.com/intern/diff/D28118182/)

[ghstack-poisoned]
